impl/parquet_reader.go

```diff
@@ -70,19 +73,50 @@ func (p *ParquetReader) GetNumberOfRows() int64 {
 }
 
 func toKafkaMessage(message ParquetMessage) (*kafka.Message, error) {
+	timestamp, err := time.Parse(time.RFC3339, message.Timestamp)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to convert string to time.Time: %s", message.Timestamp)
+	}
+
 	var headers []kafka.Header
 	if len(message.Headers) > 0 {
 		err := json.Unmarshal([]byte(message.Headers), &headers)
 		if err != nil {
 			return nil, errors.Wrapf(err, "Failed to unmarshal kafka headers: %s", message.Headers)
 		}
 	}
-	return &kafka.Message{
+
+	var timestampType int
+	switch message.TimestampType {
+	case kafka.TimestampCreateTime.String():
+		timestampType = int(kafka.TimestampCreateTime)
+	case kafka.TimestampLogAppendTime.String():
+		timestampType = int(kafka.TimestampLogAppendTime)
+	case kafka.TimestampNotAvailable.String():
+		fallthrough
+	default:
+		timestampType = int(kafka.TimestampNotAvailable)
+	}
+
+	kafkaMessage := &kafka.Message{
 		Value: []byte(message.Value),
 		TopicPartition: kafka.TopicPartition{
 			Topic: &message.Topic,
 		},
-		Key:     []byte(message.Key),
-		Headers: headers,
-	}, nil
+		Key:           []byte(message.Key),
+		Headers:       headers,
+		Timestamp:     timestamp,
+		TimestampType: kafka.TimestampType(timestampType),
+	}
+
+	if message.Offset != nil {
+		offset, err := strconv.Atoi(*message.Offset)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to convert string to int for message offset: %s", message.Offset)
```

Comment:
message.Offset is a pointer; it cannot be formatted with %s.

Reply:
message.Offset will be changed to value now, with the new approach of saving offset and partition always but optional during restore.